### ISSUE TITLE
Issue with Stream Management resumption

### ISSUE THREAD

**Issue by nesium** [strophe/libstrophe#222]
Hi there. Thanks for your work on the library!

I'm currently trying to get Stream Management to work and noticed that messages sent after a connection termination are not re-sent on reconnect. I've used a fairly simple test setup:

- Configure connection and connect to server
- Wait for successful connection then send presence stanza (and more)
- Turn off internet connection
- Send a message before TCP keepalive is sent (and before the connection is known to be terminated)
- Wait for disconnect
- Fetch SM state from current connection (and ensure that there are no more messages in the send queue), configure new connection with SM state and reconnect to server

My expectation was that the message would be re-sent automatically, but alas it was not. I could see that `<enable …/>` and `<enabled …/>` as well as `<a …/>` and `<r …/>` elements were exchanged before disconnecting.

After some debugging I've noticed that the messages in the `sm_queue` have a lower `sm_h` than the `h` received from the `<resumed h="n" …/>` element from the server. I believe that there's a problem in the `xmpp_run_once` function in the following piece of code:

```c
if (!(tsq->owner & XMPP_QUEUE_SM) && conn->sm_state->sm_enabled) {
    tsq->sm_h = conn->sm_state->sm_sent_nr;
    conn->sm_state->sm_sent_nr++;
    strophe_debug_verbose(1, ctx, "xmpp", "SM_Q_MOVE: %p", tsq);
    add_queue_back(&conn->sm_state->sm_queue, tsq);
    tsq = NULL;
}
```

Here the stanza is only pushed to the `sm_queue` - and more importantly `sm_sent_nr` only incremented - if state management is enabled (`sm_enabled`). However `sm_enabled` is only set once the `<enabled …/>` or `<resumed …/>` elements are received. So there's a potential for a race condition where the `sm_sent_nr` is untouched for stanzas sent between a `XMPP_CONN_CONNECT` event received in the connection handler and before either `<enabled …/>` or `<resumed …/>` are received. Prosody with the mod_smacks module seems to start incrementing its received stanza counter immediately after the `<enable …/>` stanza which leads to the counter mismatch.

That begs the question if this is intended behavior or an oversight. My interpretation from [the spec](https://xmpp.org/extensions/xep-0198.html) is that it should work (emphasis mine)…

> Note: Each entity maintains two counters for any given stream: a counter of stanzas it has sent, and a counter of stanzas it has received and handled ('h'). _The counter for an entity's own sent stanzas is set to zero and started after sending either `<enable/>` or `<enabled/>`_. The counter for the received stanzas ('h') is set to zero and started after receiving either `<enable/>` or `<enabled/>`.

So implementation-wise it could make sense to optimistically assume that stream management is enabled and discard the SM queue in case the server rejects the session.

Somewhat related: It would also be helpful if one could access the fields of `xmpp_sm_state_t` and construct it so that you can persist it between restarts but that's probably beyond this issue.

**Comment by sjaeckel** [comment]
Can you please try out the latest master version?

**Comment by sjaeckel** [comment]
Closing this as it's most likely a duplicate of #215 

Please re-open if this is not the case.

> Somewhat related: It would also be helpful if one could access the fields of `xmpp_sm_state_t` and construct it so that you can persist it between restarts but that's probably beyond this issue.

To make the fields public is not an option. We could introduce a serialize/deserialize API!? You're welcome to contact me via mail if you're interested in that feature.

**Comment by nesium** [comment]
Thank you so much for coming back to this, @sjaeckel! Unfortunately we've moved to a native Rust implementation in the meantime, which is why I'm not able to try the fix in #230. The switch made sense for us since we didn't have to go through an FFI layer anymore.
